src/cranelift_backend/mod.rs

```diff
@@ -165,15 +165,15 @@ impl CodeGen for hir::If {
 
 impl CodeGen for hir::Else {
     fn codegen<'a>(&'a self, context: &mut Context<'a>, builder: &mut FunctionBuilder) -> Value {
-        let expr = builder.create_block();
-        let then_values = context.eval_all_in_block(&self.expr, expr, builder);
+        let lhs = builder.create_block();
+        let then_values = context.eval_all_in_block(&self.lhs, lhs, builder);
 
         let end = context.new_block_with_arg(&self.result_type, builder);
 
         let else_none = builder.create_block();
         builder.ins().jump(else_none, &[]);
 
-        let else_values = context.eval_all_in_block(&self.otherwise, else_none, builder);
+        let else_values = context.eval_all_in_block(&self.rhs, else_none, builder);
```

Comment:
> `builder.ins().jump(else_none, &[]);`

This is an unconditional jump to the else block, but we only want to jump there conditionally if the tag of the value returned is `None`. I think an easier alternative rather than trying to get codegen working here would be desugaring the `a else b` expressions during name resolution or type checking into:
```hs
match a
| Some value -> value
| None -> b
```